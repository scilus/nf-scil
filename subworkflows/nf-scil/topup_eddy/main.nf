// ** Importing modules from nf-scil ** //
include { PREPROC_TOPUP } from '../../../modules/nf-scil/preproc/topup/main'
include { PREPROC_EDDY } from '../../../modules/nf-scil/preproc/eddy/main'
include { UTILS_EXTRACTB0 } from '../../../modules/nf-scil/utils/extractb0/main'

workflow TOPUP_EDDY {

    // ** The subworkflow will optionally run topup if a reverse b0 or reverse DWI is provided.   ** //
    // ** In both cases, it will perform EDDY and also extract a b0 from the corrected DWI image. ** //

    take:
        ch_dwi          // channel: [ val(meta), dwi, bval, bvec ]
        ch_b0           // channel: [ val(meta), b0 ]
        ch_rev_dwi      // channel: [ val(meta), rev_dwi, rev_bval, rev_bvec ]
        ch_rev_b0       // channel: [ val(meta), rev_b0Â ]
        ch_config_topup // channel

    main:
        ch_versions = Channel.empty()

        // ** Create channel for TOPUP ** //
        ch_topup = ch_dwi
            .join(ch_b0, remainder: true)
            .map{ it[0..3] + [it[4] ?: []] }

        ch_topup = ch_topup
            .join(ch_rev_dwi, remainder: true)
            .map{ it[5] ? it : it[0..4] + [[], [], []] }
            .join(ch_rev_b0, remainder: true)
            .branch{
                with_topup: it[5] || it[8]
            }

        // ** RUN TOPUP ** //
        PREPROC_TOPUP ( ch_topup.with_topup.view(), ch_config_topup.view() )
        ch_versions = ch_versions.mix(PREPROC_TOPUP.out.versions.first())

        // ** Create channel for EDDY ** //
        ch_eddy_input = ch_dwi
            .join(ch_rev_dwi, remainder: true)
            .map{ it[0..3] + [it[4] ? it[4..-1] : [], [], []] }
            .join(PREPROC_TOPUP.out.topup_corrected_b0s, remainder: true)
            .map{ it[0..6] + [it[7] ?: []] }
            .join(PREPROC_TOPUP.out.topup_fieldcoef, remainder: true)
            .map{ it[0..7] + [it[8] ?: []] }
            .join(PREPROC_TOPUP.out.topup_movpart, remainder: true)
            .map{ it[0..8] + [it[9] ?: []] }

        PREPROC_EDDY ( ch_eddy_input )
        ch_versions = ch_versions.mix(PREPROC_EDDY.out.versions.first())

        ch_dwi_extract_b0 = PREPROC_EDDY.out.dwi_corrected
            .join(PREPROC_EDDY.out.bval_corrected)
            .join(PREPROC_EDDY.out.bvec_corrected)

        UTILS_EXTRACTB0 { ch_dwi_extract_b0 }
        ch_versions = ch_versions.mix(UTILS_EXTRACTB0.out.versions.first())

    emit:
        dwi      = PREPROC_EDDY.out.dwi_corrected       // channel: [ val(meta), [ dwi_corrected ] ]
        bval     = PREPROC_EDDY.out.bval_corrected      // channel: [ val(meta), [ bval_corrected ] ]
        bvec     = PREPROC_EDDY.out.bvec_corrected      // channel: [ val(meta), [ bvec_corrected ] ]
        b0       = UTILS_EXTRACTB0.out.b0               // channel: [ val(meta), [ b0 ] ]
        b0_mask  = PREPROC_EDDY.out.b0_mask             // channel: [ val(meta), [ b0_mask ] ]
        versions = ch_versions                  // channel: [ versions.yml ]
}
