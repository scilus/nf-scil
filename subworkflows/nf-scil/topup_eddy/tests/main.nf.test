nextflow_workflow {

    name "Test Subworkflow TOPUP_EDDY"
    script "../main.nf"
    workflow "TOPUP_EDDY"
    config "./nextflow.config"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/topup_eddy"

    tag "preproc"
    tag "preproc/topup"
    tag "preproc/eddy"
    tag "utils/extractb0"

    tag "load_test_data"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "topup_eddy_light.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("eddy") {

        when {
            workflow {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.nii.gz"),
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.bval"),
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.bvec")
                    ]}
                input[1] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/sub-01_dir-AP_sbref.nii.gz")
                    ]}
                input[2] = Channel.from( [] )
                input[3] = Channel.from( [] )
                input[4] = Channel.from( [] )
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.b0_mask,
                    workflow.out.bval,
                    workflow.out.versions,
                    file(workflow.out.dwi.get(0).get(1)).name,
                    file(workflow.out.bvec.get(0).get(1)).name,
                    file(workflow.out.b0.get(0).get(1)).name).match() })
        }
    }

    test("topup_eddy_rev_b0") {
        when {
            workflow {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.nii.gz"),
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.bval"),
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.bvec")
                    ]}
                input[1] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/sub-01_dir-AP_sbref.nii.gz")
                    ]}
                input[2] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        [],
                        [],
                        []
                    ]}

                input[3] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/sub-01_dir-PA_sbref.nii.gz")
                    ]}
                input[4] = Channel.from( [] )
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.b0_mask,
                    workflow.out.bval,
                    workflow.out.versions,
                    file(workflow.out.dwi.get(0).get(1)).name,
                    file(workflow.out.bvec.get(0).get(1)).name,
                    file(workflow.out.b0.get(0).get(1)).name).match() })
        }
    }

    test("topup_eddy_rev_dwi") {

        when {
            workflow {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.nii.gz"),
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.bval"),
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.bvec")
                    ]}
                input[1] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/sub-01_dir-AP_sbref.nii.gz")
                    ]}
                input[2] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/sub-01_dir-PA_dwi.nii.gz"),
                        file("\${test_data_directory}/sub-01_dir-PA_dwi.bval"),
                        file("\${test_data_directory}/sub-01_dir-PA_dwi.bvec")
                    ]}
                input[3] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/sub-01_dir-PA_sbref.nii.gz")
                    ]}
                input[4] = Channel.from( [] )
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.b0_mask,
                    workflow.out.bval,
                    workflow.out.versions,
                    file(workflow.out.dwi.get(0).get(1)).name,
                    file(workflow.out.bvec.get(0).get(1)).name,
                    file(workflow.out.b0.get(0).get(1)).name).match() })
        }
    }
}
