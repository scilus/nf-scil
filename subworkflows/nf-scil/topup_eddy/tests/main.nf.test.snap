{
    "topup_eddy_rev_b0": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__b0_bet_mask.nii.gz:md5,3574cc8f29da429a0f610f2572c4886c"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__bval_eddy:md5,8192e97b08e8032382397bb844b31892"
                ]
            ],
            [
                "versions.yml:md5,1210f8705c10f136e83e5f07c0d77b70",
                "versions.yml:md5,dc9478b4d3cb93d93799e493aefe5220"
            ],
            "test__dwi_corrected.nii.gz",
            "test__dwi_eddy_corrected.bvec",
            "test_b0.nii.gz"
        ],
        "meta": {
            "nf-test": "0.9.0-rc1",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-06-22T14:40:30.486384"
    },
    "eddy": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__b0_bet_mask.nii.gz:md5,3574cc8f29da429a0f610f2572c4886c"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__bval_eddy:md5,8192e97b08e8032382397bb844b31892"
                ]
            ],
            [
                "versions.yml:md5,1210f8705c10f136e83e5f07c0d77b70",
                "versions.yml:md5,dc9478b4d3cb93d93799e493aefe5220"
            ],
            "test__dwi_corrected.nii.gz",
            "test__dwi_eddy_corrected.bvec",
            "test_b0.nii.gz"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-09T00:15:24.300649"
    },
    "topup_eddy_rev_dwi": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__b0_bet_mask.nii.gz:md5,3574cc8f29da429a0f610f2572c4886c"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__bval_eddy:md5,8192e97b08e8032382397bb844b31892"
                ]
            ],
            [
                "versions.yml:md5,1210f8705c10f136e83e5f07c0d77b70",
                "versions.yml:md5,dc9478b4d3cb93d93799e493aefe5220"
            ],
            "test__dwi_corrected.nii.gz",
            "test__dwi_eddy_corrected.bvec",
            "test_b0.nii.gz"
        ],
        "meta": {
            "nf-test": "0.9.0-rc1",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-06-24T05:02:10.817762"
    }
}